File: Parallel-Csharp/src/Parallel/Linq101/MyLinq/OrderbyMethods.cs
using Parallel.Linq101.DataSources;

namespace Parallel.Linq101.MyLinq;

public class OrderbyMethods {
    //original:4 target:40 form Orderings
    public List<Product> GetProductList() => Products.ProductList;
    public List<Customer> GetCustomerList() => Customers.CustomerList;

    public class CaseInsensitiveComparer: IComparer<string> {
        public int Compare(string x, string y) =>
            string.Compare(x, y, StringComparison.OrdinalIgnoreCase);
    }
    
    public class CustomNumberComparer : IComparer<int> {
        public int Compare(int x, int y) =>
            Math.Abs(x - 5) - Math.Abs(y - 5);
    }
    
    public List<double> DoubleOrderby() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        List<double> sortedNumbers = numbers.OrderBy(n => n).ToList();
        return sortedNumbers;
    }

    public List<double> DoubleOrderbyFraction() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        List<double> sortedNumbers = numbers.OrderBy(n => n-(int)n).ToList();
        return sortedNumbers;
    }
    
    public List<double> DoubleFractionOrderby() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        List<double> sortedNumbers = numbers.Select(n => n-(int)n).OrderBy(n=>n).ToList();
        return sortedNumbers;
    }
    
    public List<double> BigDoubleOrderby() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        List<double> sortedNumbers = numbers.Where(n => n > 5).OrderBy(n => n).ToList();
        return sortedNumbers;
    }

    public List<double> BigDoubleOrderbyFraction() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        List<double> sortedNumbers = numbers.Where(n => n > 5).OrderBy(n => n-(int)n).ToList();
        return sortedNumbers;
    }
    
    public List<double> BigDoubleFractionOrderby() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        List<double> sortedNumbers = numbers.Where(n => n > 5).Select(n => n-(int)n).OrderBy(n=>n).ToList();
        return sortedNumbers;
    }
    
    public List<Tuple<bool, List<double>>> DoubleGroupByBoolOrderby() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        List<Tuple<bool, List<double>>> numbersGroups = numbers.GroupBy(n => n>5)
            .Select(g => Tuple.Create(g.Key, g.OrderBy(w=>w).ToList())).ToList();
        return numbersGroups;
    }

    public Dictionary<bool, List<double>> DoubleGroupByBoolDictOrderby() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        Dictionary<bool, List<double>>  numbersGroups = numbers.GroupBy(n => n>5)
            .ToDictionary(g => g.Key, g => g.OrderBy(w=>w).ToList());
        return numbersGroups;
    }
    
    public List<Tuple<bool, List<double>>> DoubleGroupByBoolOrderbyFraction() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        List<Tuple<bool, List<double>>> numbersGroups = numbers.GroupBy(n => n>5)
            .Select(g => Tuple.Create(g.Key, g.OrderBy(n=>n-(int)n).ToList())).ToList();
        return numbersGroups;
    }

    public Dictionary<bool, List<double>> DoubleGroupByBoolDictOrderbyFraction() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        Dictionary<bool, List<double>>  numbersGroups = numbers.GroupBy(n => n>5)
            .ToDictionary(g => g.Key, g => g.OrderBy(n=>n-(int)n).ToList());
        return numbersGroups;
    }

    public List<string> StringOrderby() {
        string[] words = { "cherry", "apple", "blueberry" };
        List<string> sortedWords = words.OrderBy(w => w).ToList();
        return sortedWords;
    }

    public List<string> StringOrderbyLength() {
        string[] words = { "cherry", "apple", "blueberry" };
        List<string> sortedWords = words.OrderBy(w => w.Length).ToList();
        return sortedWords;
    }

    public List<int> StringLengthOrderby() {
        string[] words = { "cherry", "apple", "blueberry" };
        List<int> sortedWords = words.Select(w => w.Length).OrderBy(w => w).ToList();
        return sortedWords;
    }

    public List<string> ShortStringOrderby() {
        string[] words = { "cherry", "apple", "blueberry", "pear", "peach", "melon" };
        List<string> sortedWords = words.Where(w => w.Length < 6).OrderBy(w => w).ToList();
        return sortedWords;
    }

    public List<string> ShortStringOrderbyLength() {
        string[] words = { "cherry", "apple", "blueberry", "pear", "peach", "melon" };
        List<string> sortedWords = words.Where(w => w.Length < 6).OrderBy(w => w.Length).ToList();
        return sortedWords;
    }

    public List<int> ShortStringLengthOrderby() {
        string[] words = { "cherry", "apple", "blueberry", "pear", "peach", "melon" };
        List<int> sortedWords = words.Where(w => w.Length < 6).Select(w => w.Length).OrderBy(w => w).ToList();
        return sortedWords;
    }
    
    public List<Tuple<char, List<string>>> StringGroupByFirstCharOrderby() {
        string[] words = { "blueberry", "chimpanzee", "abacus", "banana", "apple", "cheese" };
        List<Tuple<char, List<string>>> wordsGroups = words.GroupBy(w => w[0])
            .Select(g => Tuple.Create(g.Key, g.OrderBy(w=>w).ToList())).ToList();
        return wordsGroups;
    }

    public Dictionary<char, List<string>> StringGroupByFirstCharDictOrderby() {
        string[] words = { "blueberry", "chimpanzee", "abacus", "banana", "apple", "cheese" };
        Dictionary<char, List<string>> wordsGroups = words.GroupBy(w => w[0])
            .ToDictionary(g => g.Key, g => g.OrderBy(w=>w).ToList());
        return wordsGroups;
    }

    public List<Tuple<bool, List<string>>> StringGroupByLengthOrderbyLength() {
        string[] words = { "blueberry", "chimpanzee", "abacus", "banana", "apple", "cheese" };
        List<Tuple<bool, List<string>>> wordsGroups = words.GroupBy(w => w.Length > 8)
            .Select(g => Tuple.Create(g.Key, g.OrderBy(w=>w.Length).ToList())).ToList();
        return wordsGroups;
    }

    public Dictionary<bool, List<string>> StringGroupByLengthDictOrderbyLength() {
        string[] words = { "blueberry", "chimpanzee", "abacus", "banana", "apple", "cheese" };
        Dictionary<bool, List<string>> wordsGroups = words.GroupBy(w => w.Length > 8)
            .ToDictionary(g => g.Key, g => g.OrderBy(w=>w.Length).ToList());
        return wordsGroups;
    }

    public List<Product> ProductOrderbyId() {
        List<Product> products = GetProductList();
        List<Product> sortedProducts = products.OrderBy(p => p.ProductId).ToList();
        return sortedProducts;
    }
    
    public List<Product> ProductOrderbyName() {
        List<Product> products = GetProductList();
        List<Product> sortedProducts = products.OrderBy(p => p.ProductName).ToList();
        return sortedProducts;
    }
    
    public List<string> ProductNameOrderby() {
        List<Product> products = GetProductList();
        List<string> sortedProducts = products.Select(p => p.ProductName).OrderBy(n=>n).ToList();
        return sortedProducts;
    }
    
    public List<Product> ExpensiveProductOrderbyId() {
        List<Product> products = GetProductList();
        List<Product> sortedProducts = products.Where(p=>p.UnitPrice>100).OrderBy(p => p.ProductId).ToList();
        return sortedProducts;
    }
    
    public List<Product> ExpensiveProductOrderbyName() {
        List<Product> products = GetProductList();
        List<Product> sortedProducts = products.Where(p=>p.UnitPrice>100).OrderBy(p => p.ProductName).ToList();
        return sortedProducts;
    }
    
    public List<string> ExpensiveProductNameOrderby() {
        List<Product> products = GetProductList();
        List<string> sortedProducts = products.Where(p=>p.UnitPrice>100).Select(p => p.ProductName).OrderBy(n=>n).ToList();
        return sortedProducts;
    }
    
    public List<Tuple<string,List<Product>>> ProductGroupByCategoryOrderbyId() {
        List<Product> products = GetProductList();
        List<Tuple<string,List<Product>>> productsGroups = products.GroupBy(p=>p.Category)
            .Select(g => Tuple.Create(g.Key,g.OrderBy(p=>p).ToList())).ToList();
        return productsGroups;
    }
    
    public Dictionary<string,List<Product>> ProductGroupByCategoryDictOrderbyId() {
        List<Product> products = GetProductList();
        Dictionary<string,List<Product>>  productsGroups = products.GroupBy(p=>p.Category)
            .ToDictionary(g => g.Key,g=>g.OrderBy(p=>p).ToList());
        return productsGroups;
    }
    public List<Tuple<string,List<Product>>> ProductGroupByCategoryOrderbyName() {
        List<Product> products = GetProductList();
        List<Tuple<string,List<Product>>> productsGroups = products.GroupBy(p=>p.Category)
            .Select(g => Tuple.Create(g.Key,g.OrderBy(p=>p.ProductName).ToList())).ToList();
        return productsGroups;
    }
    
    public Dictionary<string,List<Product>> ProductGroupByCategoryDictOrderbyName() {
        List<Product> products = GetProductList();
        Dictionary<string,List<Product>>  productsGroups = products.GroupBy(p=>p.Category)
            .ToDictionary(g => g.Key,g=>g.OrderBy(p=>p.ProductName).ToList());
        return productsGroups;
    }

     public List<string> StringOrderbyCustom() {
         string[] words = { "aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry" };
        List<string> sortedWords = words.OrderBy(w => w, new CaseInsensitiveComparer()).ToList();
        return sortedWords;
    }

    public List<string> StringOrderbyLengthCustom() {
        string[] words = { "aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry" };
        List<string> sortedWords = words.OrderBy(w => w.Length, new CustomNumberComparer()).ToList();
        return sortedWords;
    }

    public List<int> StringLengthOrderbyCustom() {
        string[] words = { "aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry" };
        List<int> sortedWords = words.Select(w => w.Length).OrderBy(w => w, new CustomNumberComparer()).ToList();
        return sortedWords;
    }
    
    public List<string> ShortStringOrderbyCustom() {
        string[] words = { "aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry" };
        List<string> sortedWords = words.Where(w => w.Length < 6).OrderBy(w => w, new CaseInsensitiveComparer()).ToList();
        return sortedWords;
    }

    public List<string> ShortStringOrderbyLengthCustom() {
        string[] words = { "aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry" };
        List<string> sortedWords = words.Where(w => w.Length < 6).OrderBy(w => w.Length, new CustomNumberComparer()).ToList();
        return sortedWords;
    }

    public List<int> ShortStringLengthOrderbyCustom() {
        string[] words = { "aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry" };
        List<int> sortedWords = words.Where(w => w.Length < 6).Select(w => w.Length).OrderBy(w => w, new CustomNumberComparer()).ToList();
        return sortedWords;
    }
    
    public List<Tuple<char, List<string>>> StringGroupByFirstCharOrderbyCustom() {
        string[] words = { "aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry" };
        List<Tuple<char, List<string>>> wordsGroups = words.GroupBy(w => char.ToLower(w[0]))
            .Select(g => Tuple.Create(g.Key, g.OrderBy(w=>w,new CaseInsensitiveComparer()).ToList())).ToList();
        return wordsGroups;
    }

    public Dictionary<char, List<string>> StringGroupByFirstCharDictOrderbyCustom() {
        string[] words = { "aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry" };
        Dictionary<char, List<string>> wordsGroups = words.GroupBy(w => char.ToLower(w[0]))
            .ToDictionary(g => g.Key, g => g.OrderBy(w=>w,new CaseInsensitiveComparer()).ToList());
        return wordsGroups;
    }

    public List<Tuple<bool, List<string>>> StringGroupByLengthOrderbyLengthCustom() {
        string[] words = { "aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry" };
        List<Tuple<bool, List<string>>> wordsGroups = words.GroupBy(w => w.Length > 8)
            .Select(g => Tuple.Create(g.Key, g.OrderBy(w=>w.Length,new CustomNumberComparer()).ToList())).ToList();
        return wordsGroups;
    }

    public Dictionary<bool, List<string>> StringGroupByLengthDictOrderbyLengthCustom() {
        string[] words = { "aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry" };
        Dictionary<bool, List<string>> wordsGroups = words.GroupBy(w => w.Length > 8)
            .ToDictionary(g => g.Key, g => g.OrderBy(w=>w.Length,new CustomNumberComparer()).ToList());
        return wordsGroups;
    }
}

File: Parallel-Java/src/parallel/linq101/mylinq/OrderbyMethods.java
package parallel.linq101.mylinq;

import org.javatuples.Pair;
import parallel.linq101.datasource.Customer;
import parallel.linq101.datasource.Data;
import parallel.linq101.datasource.Product;

import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.*;

public class OrderbyMethods {
    public List<Product> getProductList() {
        return Data.getProductList();
    }

    public List<Customer> getCustomerList() {
        return Data.getCustomerList();
    }

    public class CaseInsensitiveComparer implements Comparator<String> {
        @Override
        public int compare(String o1, String o2) {
            return o1.toLowerCase().compareTo(o2.toLowerCase());
        }
    }

    public class CustomNumberComparer implements Comparator<Integer> {
        @Override
        public int compare(Integer o1, Integer o2) {
            return Math.abs(o1 - 5) - Math.abs(o2 - 5);
        }
    }


    public List<Double> doubleOrderby() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        List<Double> sortedNumbers = Arrays.stream(numbers).boxed().sorted().toList();
        return sortedNumbers;
    }

    public List<Double> doubleOrderbyFraction() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        List<Double> sortedNumbers = Arrays.stream(numbers).boxed().sorted(Comparator.comparing(n -> n - (int) n.doubleValue())).toList();
        return sortedNumbers;
    }

    public List<Double> doubleFractionOrderby() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        List<Double> sortedNumbers = Arrays.stream(numbers).map(n -> n - (int) n).boxed().sorted().toList();
        return sortedNumbers;
    }

    public List<Double> bigDoubleOrderby() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        List<Double> sortedNumbers = Arrays.stream(numbers).filter(n -> n > 5).boxed().sorted().toList();
        return sortedNumbers;
    }

    public List<Double> bigDoubleOrderbyFraction() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        List<Double> sortedNumbers = Arrays.stream(numbers).filter(n -> n > 5).boxed().sorted(Comparator.comparing(n -> n - (int) n.doubleValue())).toList();
        return sortedNumbers;
    }

    public List<Double> bigDoubleFractionOrderby() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        List<Double> sortedNumbers = Arrays.stream(numbers).filter(n -> n > 5).map(n -> n - (int) n).boxed().sorted().toList();
        return sortedNumbers;
    }

    public List<Pair<Boolean, List<Double>>> doubleGroupByBoolOrderby() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        List<Pair<Boolean, List<Double>>> numbersGroups = Arrays.stream(numbers).boxed().collect(groupingBy(n -> n > 5))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().sorted().toList())).toList();
        return numbersGroups;
    }

    public Map<Boolean, List<Double>> doubleGroupByBoolDictOrderby() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        Map<Boolean, List<Double>> numbersGroups = Arrays.stream(numbers).boxed().collect(groupingBy(n -> n > 5))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().sorted().toList()));
        return numbersGroups;
    }

    public List<Pair<Boolean, List<Double>>> doubleGroupByBoolOrderbyFraction() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        List<Pair<Boolean, List<Double>>> numbersGroups = Arrays.stream(numbers).boxed().collect(groupingBy(n -> n > 5))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().sorted(Comparator.comparing(
                        n -> n - (int) n.doubleValue())).toList())).toList();
        return numbersGroups;
    }

    public Map<Boolean, List<Double>> doubleGroupByBoolDictOrderbyFraction() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        Map<Boolean, List<Double>> numbersGroups = Arrays.stream(numbers).boxed().collect(groupingBy(n -> n > 5))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().sorted(Comparator.comparing(
                        n -> n - (int) n.doubleValue())).toList()));
        return numbersGroups;
    }

    public List<String> stringOrderby() {
        String[] words = {"cherry", "apple", "blueberry"};
        List<String> sortedWords = Arrays.stream(words).sorted().toList();
        return sortedWords;
    }

    public List<String> stringOrderbyLength() {
        String[] words = {"cherry", "apple", "blueberry"};
        List<String> sortedWords = Arrays.stream(words).sorted(Comparator.comparing(w -> w.length())).toList();
        return sortedWords;
    }

    public List<Integer> stringLengthOrderby() {
        String[] words = {"cherry", "apple", "blueberry"};
        List<Integer> sortedWords = Arrays.stream(words).map(w -> w.length()).sorted().toList();
        return sortedWords;
    }

    public List<String> shortStringOrderby() {
        String[] words = {"cherry", "apple", "blueberry", "pear", "peach", "melon"};
        List<String> sortedWords = Arrays.stream(words).filter(w -> w.length() < 6).sorted().toList();
        return sortedWords;
    }

    public List<String> shortStringOrderbyLength() {
        String[] words = {"cherry", "apple", "blueberry", "pear", "peach", "melon"};
        List<String> sortedWords = Arrays.stream(words).filter(w -> w.length() < 6).sorted(Comparator.comparing(w -> w.length())).toList();
        return sortedWords;
    }

    public List<Integer> shortStringLengthOrderby() {
        String[] words = {"cherry", "apple", "blueberry", "pear", "peach", "melon"};
        List<Integer> sortedWords = Arrays.stream(words).filter(w -> w.length() < 6).map(w -> w.length()).sorted().toList();
        return sortedWords;
    }

    public List<Pair<Character, List<String>>> stringGroupByFirstCharOrderby() {
        String[] words = {"blueberry", "chimpanzee", "abacus", "banana", "apple", "cheese"};
        List<Pair<Character, List<String>>> wordsGroups = Arrays.stream(words).collect(groupingBy(w -> w.charAt(0)))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().sorted().toList())).toList();
        return wordsGroups;
    }

    public Map<Character, List<String>> stringGroupByFirstCharDictOrderby() {
        String[] words = {"blueberry", "chimpanzee", "abacus", "banana", "apple", "cheese"};
        Map<Character, List<String>> wordsGroups = Arrays.stream(words).collect(groupingBy(w -> w.charAt(0)))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g -> g.getValue().stream().sorted().toList()));
        return wordsGroups;
    }

    public List<Pair<Boolean, List<String>>> stringGroupByLengthOrderbyLength() {
        String[] words = {"blueberry", "chimpanzee", "abacus", "banana", "apple", "cheese"};
        List<Pair<Boolean, List<String>>> wordsGroups = Arrays.stream(words).collect(groupingBy(w -> w.length() > 8))
                .entrySet().stream().map(g -> Pair.with(g.getKey(),
                        g.getValue().stream().sorted(Comparator.comparing(w -> w.length())).toList())).toList();
        return wordsGroups;
    }

    public Map<Boolean, List<String>> stringGroupByLengthDictOrderbyLength() {
        String[] words = {"blueberry", "chimpanzee", "abacus", "banana", "apple", "cheese"};
        Map<Boolean, List<String>> wordsGroups = Arrays.stream(words).collect(groupingBy(w -> w.length() > 8))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g -> g.getValue().stream().sorted(Comparator.comparing(w -> w.length())).toList()));
        return wordsGroups;
    }

    public List<Product> productOrderbyId() {
        List<Product> products = getProductList();
        List<Product> sortedProducts = products.stream().sorted(Comparator.comparing(p -> p.getProductId())).toList();
        return sortedProducts;
    }

    public List<Product> productOrderbyName() {
        List<Product> products = getProductList();
        List<Product> sortedProducts = products.stream().sorted(Comparator.comparing(p -> p.getProductName())).toList();
        return sortedProducts;
    }

    public List<String> productNameOrderby() {
        List<Product> products = getProductList();
        List<String> sortedProducts = products.stream().map(p -> p.getProductName()).sorted().toList();
        return sortedProducts;
    }

    public List<Product> expensiveProductOrderbyId() {
        List<Product> products = getProductList();
        List<Product> sortedProducts = products.stream().filter(p -> p.getUnitPrice() > 100).sorted(Comparator.comparing(p -> p.getProductId())).toList();
        return sortedProducts;
    }

    public List<Product> expensiveProductOrderbyName() {
        List<Product> products = getProductList();
        List<Product> sortedProducts = products.stream().filter(p -> p.getUnitPrice() > 100).sorted(Comparator.comparing(p -> p.getProductName())).toList();
        return sortedProducts;
    }

    public List<String> expensiveProductNameOrderby() {
        List<Product> products = getProductList();
        List<String> sortedProducts = products.stream().filter(p -> p.getUnitPrice() > 100).map(p -> p.getProductName()).sorted().toList();
        return sortedProducts;
    }

    public List<Pair<String, List<Product>>> productGroupByCategoryOrderbyId() {
        List<Product> products = getProductList();
        List<Pair<String, List<Product>>> productsGroups = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream()
                        .sorted(Comparator.comparing(p -> p.getProductId())).toList())).toList();
        return productsGroups;
    }

    public Map<String, List<Product>> productGroupByCategoryDictOrderbyId() {
        List<Product> products = getProductList();
        Map<String, List<Product>> productsGroups = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream()
                        .sorted(Comparator.comparing(p -> p.getProductId())).toList()));
        return productsGroups;
    }

    public List<Pair<String, List<Product>>> productGroupByCategoryOrderbyName() {
        List<Product> products = getProductList();
        List<Pair<String, List<Product>>> productsGroups = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream()
                        .sorted(Comparator.comparing(p -> p.getProductName())).toList())).toList();
        return productsGroups;
    }

    public Map<String, List<Product>> productGroupByCategoryDictOrderbyName() {
        List<Product> products = getProductList();
        Map<String, List<Product>> productsGroups = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().
                        sorted(Comparator.comparing(p -> p.getProductName())).toList()));
        return productsGroups;
    }

    public List<String> stringOrderbyCustom() {
        String[] words = {"aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry"};
        List<String> sortedWords = Arrays.stream(words).sorted(new CaseInsensitiveComparer()).toList();
        return sortedWords;
    }

    public List<String> stringOrderbyLengthCustom() {
        String[] words = {"aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry"};
        List<String> sortedWords = Arrays.stream(words).sorted(Comparator.comparing(w -> w.length(),
                new CustomNumberComparer())).toList();
        return sortedWords;
    }

    public List<Integer> stringLengthOrderbyCustom() {
        String[] words = {"aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry"};
        List<Integer> sortedWords = Arrays.stream(words).map(w -> w.length()).sorted(new CustomNumberComparer()).toList();
        return sortedWords;
    }

    public List<String> shortStringOrderbyCustom() {
        String[] words = {"aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry"};
        List<String> sortedWords = Arrays.stream(words).filter(w -> w.length() < 6).sorted(new CaseInsensitiveComparer()).
                toList();
        return sortedWords;
    }

    public List<String> shortStringOrderbyLengthCustom() {
        String[] words = {"aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry"};
        List<String> sortedWords = Arrays.stream(words).filter(w -> w.length() < 6).sorted(Comparator.comparing(w -> w.length(),
                new CustomNumberComparer())).toList();
        return sortedWords;
    }

    public List<Integer> shortStringLengthOrderbyCustom() {
        String[] words = {"aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry"};
        List<Integer> sortedWords = Arrays.stream(words).filter(w -> w.length() < 6).map(w -> w.length()).
                sorted(new CustomNumberComparer()).toList();
        return sortedWords;
    }

    public List<Pair<Character, List<String>>> stringGroupByFirstCharOrderbyCustom() {
        String[] words = {"aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry"};
        List<Pair<Character, List<String>>> wordsGroups = Arrays.stream(words).collect(groupingBy(w -> Character.toLowerCase(w.charAt(0))))
                .entrySet().stream().map(g -> Pair.with(g.getKey(),
                        g.getValue().stream().sorted(new CaseInsensitiveComparer()).toList())).toList();
        return wordsGroups;
    }

    public Map<Character, List<String>> stringGroupByFirstCharDictOrderbyCustom() {
        String[] words = {"aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry"};
        Map<Character, List<String>> wordsGroups = Arrays.stream(words).collect(groupingBy(w -> Character.toLowerCase(w.charAt(0))))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g -> g.getValue().stream().sorted(new CaseInsensitiveComparer()).toList()));
        return wordsGroups;
    }

    public List<Pair<Boolean, List<String>>> stringGroupByLengthOrderbyLengthCustom() {
        String[] words = {"aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry"};
        List<Pair<Boolean, List<String>>> wordsGroups = Arrays.stream(words).collect(groupingBy(w -> w.length() > 8))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().sorted(
                        Comparator.comparing(w -> w.length(), new CustomNumberComparer())).toList())).toList();
        return wordsGroups;
    }

    public Map<Boolean, List<String>> stringGroupByLengthDictOrderbyLengthCustom() {
        String[] words = {"aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry"};
        Map<Boolean, List<String>> wordsGroups = Arrays.stream(words).collect(groupingBy(w -> w.length() > 8))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().sorted(
                        Comparator.comparing(w -> w.length(), new CustomNumberComparer())).toList()));
        return wordsGroups;
    }
}
